src/compiler/compiler.ts

```diff
@@ -0,0 +1,168 @@
+import * as ts from 'typescript';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as utils from './utils';
+
+function formatTscParserErrors(errors: ts.Diagnostic[]) {
+  return errors.map(s => JSON.stringify(s, null, 4)).join('\n');
+}
+
+interface File {
+  path: string;
+  version: number;
+  snapshot: ts.IScriptSnapshot;
+  src: string;
+}
+
+class Files {
+  private files: Map<string, File>;
+  constructor() {
+    this.files = new Map();
+  }
+  // set(path: string, file: File) {
+  //     this.files.set(path, file)
+  // }
+  update({ path, src }: { path: string; src: string }) {
+    const file = this.get(path);
+    if (file === undefined) {
+      const newFile: File = {
+        src,
+        path,
+        version: 0,
+        snapshot: ts.ScriptSnapshot.fromString(src),
+      };
+      this.files.set(path, newFile);
+      return;
+    }
+    if (file.src === src) {
+      return;
+    }
+    file.src = src;
+    file.version += 1;
+    file.snapshot = ts.ScriptSnapshot.fromString(src);
+  }
+  get(path: string) {
+    return this.files.get(path);
+  }
+  getFileNames() {
+    return Array.from(this.files.values()).map(f => f.path);
+  }
+  getScriptVersion(path: string) {
+    const f = this.get(path);
+    if (f === undefined) {
+      return '';
+    }
+    return f.version.toString();
+  }
+}
+
+export class Compiler {
+  private options?: ts.CompilerOptions;
+  private service: ts.LanguageService;
+  private files: Files;
+
+  constructor(options?: ts.CompilerOptions) {
+    this.files = new Files();
+    this.options = options;
+    this.service = this.createServiceHost();
+  }
+  setOptions(options: ts.CompilerOptions) {
+    this.options = options;
+  }
+  private createServiceHost() {
+    const that = this;
+
+    let service: ts.LanguageService;
+
+    class ServiceHost implements ts.LanguageServiceHost {
+      // getCustomTransformers() {
+      //     return {
+      //         before: [tsruntimeTransformer(service.getProgram())]
+      //     }
+      // }
+      getScriptFileNames() {
+        return that.files.getFileNames();
+      }
+      getScriptVersion(fileName: string) {
+        return that.files.getScriptVersion(fileName);
+      }
+      getScriptSnapshot(fileName: string) {
+        const file = that.files.get(fileName);
+        if (file !== undefined) {
+          return file.snapshot;
+        }
+        if (!fs.existsSync(fileName)) {
```

Comment:
I am not sure how bad this would be for large apps with loads of files.
But maybe in order to avoid an extra IO operation you can do

```ts
try {
  return ts.ScriptSnapshot.fromString(fs.readFileSync(fileName).toString()); 
} catch {
 return undefined;
}

```